Readme.md
# mock-response [![NPM version][npm-image]][npm-url] [![build status][travis-image]][travis-url] [![Test coverage][coveralls-image]][coveralls-url]

> Response renderer

## Installation

    npm install mock-response

## Usage

    var Response = require('mock-response');
    var response = new Response('./test/fixtures/deal/26701138.GET.json');
    // response.status => 200
    // response.body =>
    // {
    //     "hello": "world"
    // }


For more info, see [docs](browse/docs/Home.md).

## License

MIT

[npm-image]: https://img.shields.io/npm/v/mock-response.svg?style=flat
[npm-url]: https://npmjs.org/package/mock-response
[travis-image]: https://img.shields.io/travis/meituan/mock-response.svg?style=flat
[travis-url]: https://travis-ci.org/meituan/mock-response
[coveralls-image]: https://img.shields.io/coveralls/meituan/mock-response.svg?style=flat
[coveralls-url]: https://coveralls.io/r/meituan/mock-response?branch=master


History.md
3.0.0 / 2015-01-26
==================

  * Rename params to overrides
  * Rename wildcards to params

2.0.0 / 2015-01-26
==================

  * Migrate `atom` and `mixin` inside mock-response

1.0.2 / 2015-01-23
==================

  * Migrate params extend into mock, because extension should after atom and mixin

1.0.1 / 2015-01-23
==================

  * Fix magic params override entire object

1.0.0 / 2015-01-20
==================

  * Seperate mock-response from mock


test/utility.js
var utility = require('../lib/utility');

describe('utility', function() {
    describe('walk', function() {
        it('should only execute callback with primitive value', function() {
            utility.walk({
                arr: ['a', 'b'],
                obj: {
                    a: 'a',
                    b: 'b',
                }
            }, function(value) {
                value.should.be.type('string');
            });
        });
    });

    describe('set', function() {
        it('should handle nested object', function() {
            var object = {
                a: {
                    b: {
                        c: 0
                    }
                }
            };
            utility.set(object, 'a.b.c', 1);
            object.a.b.c.should.equal(1);
        });

        it('should handle array key index', function() {
            var object = {
                a: [{
                    b: 0
                }, {
                    c: 1
                }]
            };
            utility.set(object, 'a.0.b', 2);
            object.a[0].b.should.equal(2);
            object.a[1].c.should.equal(1);
        });

        it('should handle property don\'t exist', function() {
            var object = {};
            utility.set(object, 'a.b.c', 1);
            object.a.b.c.should.equal(1);
        });
    });
});


test/response.js
var fs = require('fs');
var Response = require('..');

describe('Response', function() {
    var mount = './test/fixtures';

    it('should allow mock data with out meta data', function() {
        var res = new Response('shop/:id.GET.json', {
            mount: mount,
        });
        res.status.should.equal(200);
        res.body.should.be.type('object');
        res.body.shopId.should.be.ok;
    });

    it('should read the content of JSON response', function() {
        var res = new Response('deal/26701138.GET.json', {
            mount: mount,
        });
        res.status.should.equal(200);
        res.body.should.be.type('object');
    });

    it('should always use newest version of JSON response', function() {
        var res = new Response('cache.json', {
            mount: mount,
        });
        res.status.should.equal(200);
        res.body.hello.should.eql('world');
        // change content of the file
        fs.writeFileSync('test/fixtures/cache.json', '{"hello":"world2"}', 'utf8');
        res = new Response('cache.json', {
            mount: mount,
        });
        res.status.should.equal(200);
        res.body.hello.should.eql('world2');
        // restore change
        fs.writeFileSync('test/fixtures/cache.json', '{"hello":"world"}\n', 'utf8');

        // Nodejs module
        res = new Response('cache.js', {
            mount: mount,
        });
        res.status.should.equal(200);
        res.body.hello.should.eql('world');
        // change content of the file
        fs.writeFileSync('test/fixtures/cache.js', 'module.exports = {"hello":"world2"};', 'utf8');
        res = new Response('cache.js', {
            mount: mount,
        });
        res.status.should.equal(200);
        res.body.hello.should.eql('world2');
        // restore change
        fs.writeFileSync('test/fixtures/cache.js', 'module.exports = {"hello":"world"};\n', 'utf8');
    });

    it('should allow data simulation with MockJS', function() {
        var res = new Response('response/mockjs.js', {
            mount: mount,
            params: {param1: 'world'},
        });
        res.body.function.should.equal(res.body.title);
        res.body.param.should.equal('world');
    });

    it('should parse handlebar template', function() {
        var res = new Response('response/template.hbs', {
            mount: mount,
            params: {param1: 'world'},
        });
        res.body.hello.should.equal('world');
    });

    it('should parse Node.js module', function() {
        var res = new Response('response/module.js', {
            mount: mount,
        });
        res.body.hello.should.equal('world');
    });

    it('should handle Node.js module as function', function() {
        var res = new Response('response/function.js', {
            mount: mount,
            params: {param1: 'world'},
        });
        res.body.hello.should.equal('world');
    });

    it('should handle response exception', function() {
        var res = new Response('response/exception.json', {
            mount: mount,
        });
        res.status.should.equal(500);
        res.body.message.should.ok;
    });

    describe('wildcard', function() {
        it('should extract data from wildcard match', function() {
            var res = new Response('shop/:id.GET.json', {
                mount: mount,
            });
            res.body.shopId.should.equal(123456);
        });
    });

    describe('overrides', function() {
        it('should merge additional magic querystring', function() {
            var res = new Response('overrides.json', {
                mount: mount,
                overrides: {'a.1.c': 'world2'},
            });
            res.body.a[0].b.should.equal('hello');
            res.body.a[1].c.should.equal('world2');
        });

        it('should override after random data generated', function() {
            var res = new Response('overrides.json', {
                mount: mount,
                overrides: {'random.1.hello': 'world2'},
            });
            res.body.random[0].hello.should.equal('world');
            res.body.random[1].hello.should.equal('world2');
        });

        it('should work after atom expandsion', function() {
            var res = new Response('atom.json', {
                mount: mount,
                overrides: {'deal.price': 300},
            });
            res.body.deal.price.should.equal(300);
        });
    });

    describe('mixin', function() {
        it('should mixin parent dataset', function() {
            var res = new Response('mixin.json', {
                mount: mount,
            });
            res.mixin.should.eql('base.json');
            res.body.should.eql({
                base: 'base',
                child: 'child',
            });
        });

        it('should handle multiple mixins', function() {
            var res = new Response('multi-mixin.json', {
                mount: mount,
            });
            res.mixin.should.eql(['base.json', 'head.json']);
            res.body.should.eql({
                base: 'base',
                head: 'head',
                child: 'child',
            });
        });

        it('should override mixin property', function() {
            var res = new Response('override-mixin.json', {
                mount: mount,
            });
            res.mixin.should.eql('base.json');
            res.body.should.eql({
                base: 'not base',
            });
        });
    });

    describe('atom', function() {
        it('should recursively expand atom', function() {
            var res = new Response('atom.json', {
                mount: mount,
            });
            res.status = 200;
            res.body.should.eql({
                deal: {
                    id: '123456',
                    price: 200,
                }
            });
        });
    });
});


test/fixtures/cache.js
module.exports = {"hello":"world"};


lib/utility.js
var _ = require('lodash');

/**
 * Walk through object, execute callback on every primtive types.
 * @param {Object} object
 * @param {Function} callback
 * @param {Object} thisArg
 */
exports.walk = function(object, callback, thisArg) {
    _.forEach(object, function(value, key) {
        if (_.isObject(object[key])) {
            exports.walk(object[key], callback, thisArg);
        } else {
            // primtive types
            callback.call(thisArg, value, key, object);
        }
    });
};

/**
 * Exampe:
 *     var object = {
 *         a: {
 *             b: [0]
 *         }
 *     };
 *     set(object, 'a.b.0', 1);
 *     =>
 *     object.a.b === 1
 * @param {Object} object
 * @param {string} keys
 * @param {*} value
 */
exports.set = function(object, keys, value) {
    keys = keys.split('.');
    keys.reduce(function(obj, key, index) {
        if (index === keys.length - 1) {
            // last element
            obj[key] = value;
            return;
        }

        return obj[key] || (obj[key] = {});
    }, object);
    return object;
};


lib/response.js
var fs = require('fs');
var path = require('path');
var _ = require('lodash');
var handlebars = require('handlebars');
var generator = require('mock-generator');
var Utility = require('./utility');

/**
 * @constructor
 * @param {string} locator Match file path
 * @param {Object} options
 *     {
 *         params: Object
 *         overrides: Object
 *     }
 * @return
 *     {
 *         status: number
 *         body: Object
 *     }
 */  
function Response(location, options) {
    this.options = options || {};
    this.options.mount = this.options.mount || '';
    this.options.overrides = this.options.overrides || {};
    this.options.params = this.options.params || {};

    this.read(path.join(this.options.mount, location));

    ['mixin', 'atom', 'override'].forEach(function(func) {
        this[func]();
    }, this);

    return this.data;
}

Response.EXTENSIONS = ['json', 'js', 'hbs'];

/**
 * Read the content of file and parse it into json.
 * @param {string} location
 * @return {Object}
 */
Response.prototype.read = function(location) {
    var json, extension;

    extension = path.extname(location).slice(1);

    try {
        if (Response.EXTENSIONS.indexOf(extension) !== -1) {
            json = this[extension](location);
        } else {
            // treate mock file as json by default
            json = this.json(location);
        }
    } catch (err) {
        json = {
            status: 500,
            body: {
                message: 'Parse error',
                stack: err.stack,
            },
        };
    } finally {
        // prevent atom or mixin data override
        json = _.clone(json);

        if (json.body) {
            this.data = json;
        } else {
            this.data = {
                status: 200,
                body: json,
            };
        }

        // Merge wildcard match and magic `mk_` querystring into response.
        _.forOwn(this.options.params, function(value, key) {
            if (this.data.body[key] === undefined) {
                this.data.body[key] = value;
            }
        }, this);
        // Expand random data with mock-random.
        this.data.body = generator(this.data.body);
    }
};

Response.prototype.json = function(location) {
   return JSON.parse(fs.readFileSync(location, 'utf8'));
};

Response.prototype.hbs = function(location) {
    var source = fs.readFileSync(location, 'utf8');
    var data = _.merge({}, this.options.params, this.options.overrides);
    var json = handlebars.compile(source)(data);
    return JSON.parse(json);
};

Response.prototype.js = function(location) {
    location = path.join(process.cwd(), location);
    var mod = require(location);
    // remove cache
    delete require.cache[require.resolve(location)];
    if (typeof mod === 'function') {
        return mod(_.merge({}, this.options.params, this.options.overrides));
    }
    return mod;
};

/**
 * Expand mixin and atom to a response object.
 */
Response.prototype.expand = function(location) {
    return new Response(location, {
        mount: this.options.mount
    });
};

/**
 * Recursively mix body data from other dataset.
 */
Response.prototype.mixin = function() {
    var data = this.data;
    var mixin, mixins;
    if (data.mixin) {
        if (typeof data.mixin === 'string') {
            mixins = [data.mixin];
        } else {
            mixins = data.mixin.slice();
        }
        while (mixin = mixins.shift()) {
            // recursively merge parent dataset
            mixin = this.expand(mixin);
            // only affect body data
            data.body = _.assign(mixin.body, data.body);
        }
    }
};

/**
 * Recursively expand data entry with atom data.
 */
Response.prototype.atom = function() {
    Utility.walk(this.data.body, function(value, key, object) {
        if (typeof value === 'string' && value.indexOf('atom://') === 0) {
            var atom = this.expand(value.slice(7));
            // only merge atom body
            object[key] = atom.body;
        }
    }, this);
};

/**
 * Override response field by magic overrides.
 */
Response.prototype.override = function() {
    _.forOwn(this.options.overrides, function(value, key) {
        Utility.set(this.data.body, key, value);
    }, this);
};

module.exports = Response;
